Adding support for parsing application metadata

mprashant24 opened the pull request.

mprashant24 commented:
Renamed the PersonOfContactMetadata to PointOfContactMetadata

gmoore commented on src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java:
```diff
@@ -0,0 +1,48 @@
+package com.contrastsecurity.utils;
```
This is a pretty good size piece of code. Can you add a few tests around it?

mprashant24 replied:
Added test cases for Metadata deserializer.

gmoore commented:
@mprashant24 This is working well for getting a single application, but getting all applications is not retrieving the metadata. For example, this block of code prints nothing:

```            
            Applications apps = contrastSDK.getApplications(orgUuid);
            for (Application app : apps.getApplications()) {
                System.out.println(app.getName());

                if(app.getMetadataEntities() != null) {
                    List<MetadataEntity> entities = Arrays.asList(app.getMetadataEntities());
                    for(MetadataEntity entity : entities) {
                        System.out.println(entity);
                    }
                }
            }
```
It's weird to have metadata returned in the application entity call but not the applications entity call

Can you add `&expand=metadata` to UrlBuilder.getApplicationsUrl and update any associated tests? I think that'll do the trick.

mprashant24 commented on src/test/java/com/contrastsecurity/ContrastSDKTest.java:
```diff
@@ -1,158 +1,188 @@
-package com.contrastsecurity;
```
May be auto formatting.

gmoore replied:
Looking a lot better now. Thanks.

mprashant24 commented:
@gmoore I dont see metadata in applications endpoint 
![image](https://user-images.githubusercontent.com/7561057/68325167-e4169100-0096-11ea-9b38-0f7a1457a019.png)

mprashant24 commented:
@gmoore I dont see metadata in applications endpoint
![image](https://user-images.githubusercontent.com/7561057/68325235-ff819c00-0096-11ea-9b9a-c6cec34975db.png)

gmoore commented:
It works, it's just not in the docs for some reason. Regardless, we don't pass any expands in the applications list so no need to start now. I've always found that a bit awkward. 

When you sort out the problem with the entire SDKTest showing up as changed, then I can merge.

mprashant24 commented:
@gmoore I have found the issue it was due to my git version, I had a git version which had core.autocurl true. I updated my git and reverted the changes to HEAD-1 and committed my changes again. Please diff it with HEAD^^ for ContrastSDKTest and you will see only new changes.

mprashant24 commented:
![image](https://user-images.githubusercontent.com/7561057/68430728-9a08da80-017e-11ea-9ae8-f53de48af6c9.png)

gmoore commented:
@mprashant24 Did you mean to add 1dd0616 to this PR?

mprashant24 commented:
@gmoore Yes, those changes are added for customer reporting.

gmoore commented:
It looks like that commit doesn't apply to application metadata. I'd like to approve and merge the application metadata changes before reviewing new work.

mprashant24 commented:
@gmoore reverted. please close this PR so I can open another for yearly trend filtered by application.

gmoore commented:
@mprashant24 Will do but the CI build is failing: https://travis-ci.org/Contrast-Security-OSS/contrast-sdk-java/builds/610951059?utm_source=github_status&utm_medium=notification

mprashant24 commented:
@gmoore just saw that, I believe I reverted SDKTest too. Fixed and pushed the new change.

gmoore commented:
@mprashant24 Looking good. Tested this today and it's working well.

Looks like you missed a few spots for person->point conversion. I'll mark them in a new review. When those are fixed, I'll merge and release.

gmoore commented on src/main/java/com/contrastsecurity/models/MetadataEntity.java:
```diff
@@ -0,0 +1,52 @@
+            throw new InvalidConversionException(type.toString(), MetadataType.NUMERIC.toString());
+    }
+
+    public PointOfContactMetadata getAsPersonOfContactMetadata() throws InvalidConversionException{
```
Rename to `getAsPointOfContactMetadata`

gmoore commented on src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java:
```diff
@@ -0,0 +1,48 @@
+                case "NUMERIC":
+                    return context.deserialize(jsonObject,
+                            NumericMetadata.class);
+                case "PERSON_OF_CONTACT":
```
Don't change this one as it is the API key sent in JSON

gmoore commented on src/main/java/com/contrastsecurity/models/MetadataEntity.java:
```diff
@@ -0,0 +1,52 @@
+public abstract class MetadataEntity {
+
+    public enum MetadataType {
+        STRING, NUMERIC, PERSON_OF_CONTACT;
```
POINT_OF_CONTACT

gmoore commented:
Closing this. I'll just finish it here: https://github.com/Contrast-Security-OSS/contrast-sdk-java/pull/35
